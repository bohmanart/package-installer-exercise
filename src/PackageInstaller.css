.PackageInstaller {
    margin: 2rem;

}
