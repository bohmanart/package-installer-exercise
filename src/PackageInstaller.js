import React, {Component} from 'react'
import PropTypes from 'prop-types'
import PackageInput from './components/PackageInput/PackageInput'
import PackageOutput from './components/PackageOutput/PackageOutput'

import './PackageInstaller.css'

export default class PackageInstaller extends Component {
    static propTypes = {
        input: PropTypes.array.isRequired
    }

    static defaultProps = {
        dependencyCycleError: 'ERROR: dependency specification contains cycle and is therefore invalid',
        dependencyAbsentError: 'ERROR: dependency specification contains a dependency that does not exist and is therefore invalid'
    }
    
    state = {
        input: this.props.input,
        output: []
    }

    componentDidMount() {
        this.getPackageInstallOrder(this.state.input)
    }
    
    render() {
        return (
            <div className="PackageInstaller">
                <h1>Package Installer</h1>
                <p>A package installer takes a list of packages with dependencies, and installs the packages in order such that an install wonâ€™t fail due to a missing dependency. This project represents code that will determine the order of install.</p>
                <PackageInput input={this.state.input.join('", \n  "')}/>
                <PackageOutput output={this.state.output}/>
            </div>
        )
    }
    
    getPackageInstallOrder(arr) {
        let output = this.dfsTopoSort(arr)

        this.setState({output}, () => {
            this.errorChecker(output)
        })
    }

    dfsTopoSort(pkgsArr) {
        let packages = []
        let visited = []
        let sorted = []
        let hasCycleDependency = false
        let hasAbsentDependency = false

        pkgsArr.forEach(pkgStr => {
            const packageObj = this.convertPackage(pkgStr)

            packages[packageObj.packageName] = packageObj
        })

        pkgsArr.forEach(pkgStr => {
            const pkgObj = this.convertPackage(pkgStr)
            const {packageName} = pkgObj

            if (!visited.includes(packageName)) {
                visit(pkgObj)
            }
        })

        function visit(pkgObj) {
            const {packageName, packageDependency} = pkgObj

            visited.push(packageName)

            if (packageDependency) {
                const dependencyPackageObj = packages[packageDependency]

                if (dependencyPackageObj && !visited.includes(packageDependency)) {
                    // console.log('RECURSE')
                    visit(dependencyPackageObj)
                }

                if (dependencyPackageObj && !sorted.includes(packageDependency)) {
                    hasCycleDependency = true
                }

                if (!dependencyPackageObj) {
                    hasAbsentDependency = true
                }

            }

            sorted.push(packageName)
        }

        if (hasCycleDependency) {
            return this.props.dependencyCycleError
        } else if (hasAbsentDependency) {
            return this.props.dependencyAbsentError
        } else {
            return sorted.join(', ')
        }
    }

    convertPackage(pkgStr) {
        const packageArray = pkgStr.split(': ')

        return {
            packageName: packageArray[0],
            packageDependency: packageArray[1]
        }
    }

    errorChecker(output) {
        if (output === this.props.dependencyCycleError) {
            throw this.props.dependencyCycleError
        }

        if (output === this.props.dependencyAbsentError) {
            throw this.props.dependencyAbsentError
        }
    }

}
