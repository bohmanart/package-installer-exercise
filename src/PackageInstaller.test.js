import React from 'react'
import ReactDOM from 'react-dom'
import {shallow, mount} from 'enzyme'
import toJSON from 'enzyme-to-json'
import PackageInstaller from './PackageInstaller'

describe('<PackageInstaller/>', () => {
    
    it('should accept an array of strings', () => {
        const input = ['a: b', 'b: ']

        expect(new PackageInstaller({input}).props.input).toBe(input)
    })

    describe('should render', () => {
        let input

        beforeEach(() => {
            input = [
                'KittenService: ',
                'Leetmeme: Cyberportal',
                'Cyberportal: Ice',
                'CamelCaser: KittenService',
                'Fraudstream: Leetmeme',
                'Ice: '
            ]
        })

        test('shallow as expected', () => {
            const component = shallow(<PackageInstaller input={input}/>)
            const tree = toJSON(component)

            expect(tree).toMatchSnapshot()
        })

        test('mount as expected', () => {
            const component = mount(<PackageInstaller input={input}/>)
            const tree = toJSON(component)

            expect(tree).toMatchSnapshot()
        })

    })

    describe('should reject', () => {

        test('if a dependency does not exist', () => {
            const input = [
                'KittenService: ',
                'Leetmeme: Cyberportal',
                'Cyberportal: Ice',
                'CamelCaser: KittenService',
                'Fraudstream: Leetmeme',
                'Ice: LoxodontaAfricana'
            ]

            const component = shallow(<PackageInstaller input={input}/>)
            const dfsTopoSort = component.instance().dfsTopoSort(input)
            const actual = () => component.instance().errorChecker(dfsTopoSort)
            const expected = 'ERROR: dependency specification contains a dependency that does not exist and is therefore invalid'

            expect(actual).toThrowError(expected)
        })

        test('if a cycle exists', () => {
            const input = [
                'KittenService: ',
                'Leetmeme: Cyberportal',
                'Cyberportal: Ice',
                'CamelCaser: KittenService',
                'Fraudstream: ',
                'Ice: Leetmeme'
            ]

            const component = shallow(<PackageInstaller input={input}/>)
            const dfsTopoSort = component.instance().dfsTopoSort(input)
            const actual = () => component.instance().errorChecker(dfsTopoSort)
            const expected = 'ERROR: dependency specification contains cycle and is therefore invalid'

            expect(actual).toThrowError(expected)
        })

    })


    it('should output a comma seperated string of package names', () => {
        const input = [
            'KittenService: ',
            'Leetmeme: Cyberportal',
            'Cyberportal: Ice',
            'CamelCaser: KittenService',
            'Fraudstream: Leetmeme',
            'Ice: '
        ]

        const component = shallow(<PackageInstaller input={input}/>)
        const actual = component.instance().dfsTopoSort(input)
        const expected = 'KittenService, Ice, Cyberportal, Leetmeme, CamelCaser, Fraudstream'

        expect(actual).toEqual(expected)
    })

})
