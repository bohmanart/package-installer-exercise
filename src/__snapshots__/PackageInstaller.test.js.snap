// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackageInstaller/> should render mount as expected 1`] = `
<PackageInstaller
  dependencyAbsentError="ERROR: dependency specification contains a dependency that does not exist and is therefore invalid"
  dependencyCycleError="ERROR: dependency specification contains cycle and is therefore invalid"
  input={
    Array [
      "KittenService: ",
      "Leetmeme: Cyberportal",
      "Cyberportal: Ice",
      "CamelCaser: KittenService",
      "Fraudstream: Leetmeme",
      "Ice: ",
    ]
  }
>
  <div
    className="PackageInstaller"
  >
    <h1>
      Package Installer
    </h1>
    <p>
      A package installer takes a list of packages with dependencies, and installs the packages in order such that an install won’t fail due to a missing dependency. This project represents code that will determine the order of install.
    </p>
    <PackageInput
      input="KittenService: \\", 
      \\"Leetmeme: Cyberportal\\", 
      \\"Cyberportal: Ice\\", 
      \\"CamelCaser: KittenService\\", 
      \\"Fraudstream: Leetmeme\\", 
      \\"Ice: "
    >
      <div>
        <h2>
          Input
        </h2>
        <pre
          className="PackageInput"
        >
          <code>
            [
            <br />
              "
            KittenService: ", 
              "Leetmeme: Cyberportal", 
              "Cyberportal: Ice", 
              "CamelCaser: KittenService", 
              "Fraudstream: Leetmeme", 
              "Ice: 
            "
            <br />
            ]
          </code>
        </pre>
      </div>
    </PackageInput>
    <PackageOutput
      output="KittenService, Ice, Cyberportal, Leetmeme, CamelCaser, Fraudstream"
    >
      <div>
        <h2>
          Output
        </h2>
        <pre
          className="PackageOutput"
        >
          <samp>
            "
            KittenService, Ice, Cyberportal, Leetmeme, CamelCaser, Fraudstream
            "
          </samp>
        </pre>
      </div>
    </PackageOutput>
  </div>
</PackageInstaller>
`;

exports[`<PackageInstaller/> should render shallow as expected 1`] = `
<div
  className="PackageInstaller"
>
  <h1>
    Package Installer
  </h1>
  <p>
    A package installer takes a list of packages with dependencies, and installs the packages in order such that an install won’t fail due to a missing dependency. This project represents code that will determine the order of install.
  </p>
  <PackageInput
    input="KittenService: \\", 
    \\"Leetmeme: Cyberportal\\", 
    \\"Cyberportal: Ice\\", 
    \\"CamelCaser: KittenService\\", 
    \\"Fraudstream: Leetmeme\\", 
    \\"Ice: "
  />
  <PackageOutput
    output={Array []}
  />
</div>
`;
